File: Data_Structure/DS_Using_Python/Implement_Queue_Using_LinkedList.py
class Node:
	def __init__(self, data):
		self.data = data
		self.next = None
		
class Queue:
	def __init__(self):
		self.front = None
		self.rear = None
		
	def enQueue(self, item):
		ptr = Node(item)
		print(item, end=' ')
		if self.front is None:
			self.front = ptr
			self.rear = ptr
		else:
			self.rear.next = ptr
			self.rear = ptr
			
	def deQueue(self):
		if self.front is None:
			print('Queue Underflow !!')
			return
		temp = self.front
		print('\n\nDeleted item is {}'.format(temp.data))
		self.front = self.front.next
		if self.front is None:
			self.rear = None
			
		item = temp.data
		return item
		
	def peek(self):
		if self.front:
			print('\n\nFront element of queue is {}'.format(self.front.data))
		else:
			return
			
	def isEmpty(self):
		return self.front is None and self.rear is None
		
	def linkedListTraversal(self):
		temp = self.front
		while temp is not None:
			print(temp.data, end = ' ')
			temp = temp.next
			
		
q = Queue()

print('Queue is')
q.enQueue(61)
q.enQueue(11)
q.enQueue(52)
q.enQueue(45)
q.enQueue(9)

print("\n\nQueue before removing element is:")
q.linkedListTraversal()

q.deQueue()

print("\nQueue after removing element is:")
q.linkedListTraversal()

q.peek()

if(q.isEmpty()):
    print("\nQueue is empty")
else:
    print("\nQueue is not empty")

File: Data_Structure/Queue/Implement_Queue_Using_LinkedList.c
#include<stdio.h>
#include<stdlib.h>

struct node
{
    int data;
    struct node *next;
};

struct node *front = NULL;
struct node *rear = NULL;

struct node *newNode(int item)
{
    struct node *ptr = (struct node*)malloc(sizeof(struct node));
    if(ptr != NULL)
    {
        ptr -> data = item;
        ptr -> next = NULL;
        return ptr;
    }
    else
    {
        printf("Stack Overflow !\n");
    }
}

int deQueue()
{
    if(front == NULL)
    {
        printf("Queue Underflow !\n");
    }
    struct node *temp = front;
    printf("\n\nDeleted item is: %d\n",temp -> data);
    front = front -> next;
    if(front == NULL)
    {
        rear = NULL;
    }
    int item = temp -> data;
    free(temp);
    return item;
}

void enQueue(int item)
{
    struct node* ptr = newNode(item);
    printf("%d ", item);
    if(front == NULL)
    {
        front = ptr;
        rear = ptr;
    }
    else
    {
        rear -> next = ptr;
        rear = ptr;
    }
}

void linkedListTraversal()
{
    struct node *temp = front;
    while(temp != NULL)
    {
        printf("%d ",temp -> data);
        temp = temp -> next;
    } 
}

void peek()
{
    if(front != NULL)
    {
        printf("\n\nFront element is %d",front -> data);
    }
    else
    {
        return;
    }
}

int isEmpty()
{
    return front == NULL && rear == NULL;
}

int main()
{
    printf("Queue is:\n");
    enQueue(51);
    enQueue(6);
    enQueue(14);
    enQueue(22);
    enQueue(63);
    
    peek();
    
    printf("\n\nQueue before removing element is:\n");
    linkedListTraversal();
    
    deQueue();
    
    if(isEmpty())
    {
        printf("\nQueue is empty");
    }
    else
    {
        printf("\nQueue is not empty");
    }
    
    printf("\n\nQueue after removing element is:\n");
    linkedListTraversal();
    return 0;
}